Code/src/main/java/nl/utwente/group10/ui/components/blocks/Block.java

```diff
@@ -31,124 +44,258 @@
  * Each block implementation should also feature it's own FXML implementation.
  * </p>
  */
-public abstract class Block extends StackPane implements ComponentLoader {
-
+@SuppressWarnings("restriction")
+public abstract class Block extends StackPane implements ComponentLoader, ConnectionStateDependent, VisualStateDependent {
     /** The pane that is used to hold state and place all components on. */
     private CustomUIPane parentPane;
-    /** The context menu associated with this block instance. */
+    
+    /** The Circle (Context) menu associated with this block instance. */
     private CircleMenu circleMenu;
     
-    /** The connection state this Block is in */
-    private int connectionState;
+    /** The expression of this Block. */
+    protected Expr expr;
+    
+    /** Property for the ConnectionState. */
+    protected IntegerProperty connectionState;
+    
+    /** Property for the VisualState. */
+    protected IntegerProperty visualState;
+    
+    /** Property for the expression freshness. */
+    private BooleanProperty exprIsDirty;
 
     /**
+     * Constructs a new Block.
      * @param pane
      *            The pane this block belongs to.
      */
     public Block(CustomUIPane pane) {
         parentPane = pane;
-
-        parentPane.selectedBlockProperty()
-                .addListener(
-                        event -> {
-                            if (parentPane.getSelectedBlock().isPresent()
-                                    && this.equals(parentPane
-                                            .getSelectedBlock().get())) {
-                                this.getStyleClass().add("selected");
-                            } else {
-                                this.getStyleClass().removeAll("selected");
-                            }
-                        });
-
+        connectionState = new SimpleIntegerProperty(ConnectionCreationManager.getConnectionState());
+        visualState = new SimpleIntegerProperty(ConnectionCreationManager.getConnectionState());
+        exprIsDirty = new SimpleBooleanProperty(true);
+        
+        // Add listeners to the states.
+        // Invalidate listeners give the Block a change to react on the state
+        // change before it is cascaded.
+        // Cascade listeners make sure state changes are correctly propagated.
+        connectionState.addListener(a -> invalidateConnectionState());
+        connectionState.addListener(a -> setExprIsDirty(true));
+        connectionState.addListener(this::cascadeConnectionState);
```

Comment:
Still has a different effect then. On top of that, cascadeConnectionState uses the new value (which would then have to be gotten using a getConnectionState() call.